tao_compiler/mlir/disc/transforms/lhlo_legalize_roots_to_loops.cc

```diff
@@ -4098,6 +4098,248 @@ LogicalResult lowerWithScheduleSparseFillEmptyRowsOpCPU(
   return success();
 }
 
+LogicalResult lowerWithScheduleSparseSegmentMeanOpCPU(
+    ArrayRef<Operation*> root_ops, Operation* dominant_op,
+    Block* parent = nullptr, bool non_fusion = false,
+    const ShapeAnalysis* shape_analysis = nullptr) {
+  if (!(root_ops.size() == 1 &&
+        isa<lmhlo_disc::SparseSegmentMeanOp>(root_ops[0]))) {
+    return dominant_op->emitError()
+           << "root_ops[0] is not a lmhlo_disc::SparseSegmentMeanOp";
+  }
+  auto sparse_segment_mean =
+      dyn_cast<lmhlo_disc::SparseSegmentMeanOp>(root_ops[0]);
+  if (!sparse_segment_mean) {
+    return dominant_op->emitError()
+           << "can not cast root_ops[0] to lmhlo_disc::SparseSegmentMeanOp";
+  }
+
+  auto loc = sparse_segment_mean.getLoc();
+  OpBuilder b(root_ops.back());
+
+  b.setInsertionPoint(sparse_segment_mean.getOperation());
+
+  // input
+  Value input = sparse_segment_mean.getData();
+  Value indices = sparse_segment_mean.getIndices();
+  Value segment_ids = sparse_segment_mean.getSegmentIds();
+  // output
+  Value output = sparse_segment_mean.getOutput();
+  MemRefType output_type = output.getType().template cast<MemRefType>();
+  int64_t output_rank = output_type.getRank();
+
+  Value zero = b.create<arith::ConstantIndexOp>(loc, 0);
+  Value one = b.create<arith::ConstantIndexOp>(loc, 1);
+
+  Value zero_floating = b.create<arith::ConstantOp>(
+      loc, b.getFloatAttr(output_type.getElementType(), 0));
+  Value one_floating = b.create<arith::ConstantOp>(
+      loc, b.getFloatAttr(output_type.getElementType(), 1));
+
+  Value num_results = b.create<memref::DimOp>(loc, output, 0);
+
+  Value segment_count_memref;
+  // segment_count[max_segment_ids]
+  // for (i = 0; i < segment_ids.shape(0); ++i) {
+  //   segment_idx = segment_ids[i]
+  //   segment_count[segment_idx]++
+  // }
+  {
+    auto alloc = b.create<memref::AllocOp>(
+        loc,
+        MemRefType::get({-1}, output_type.getElementType(),
+                        MemRefLayoutAttrInterface(),
+                        StringAttr::get(sparse_segment_mean->getContext(),
+                                        placement_utils::kCpu)),
+        num_results);
+    segment_count_memref = alloc.getResult();
+
+    {
+      auto for_op =
+          b.create<scf::ForOp>(loc, /* lowerBound */ zero,
+                               /* upperBound */ num_results, /* step */ one);
+      for_op.getBody()->clear();
+      b.setInsertionPointToStart(for_op.getBody());
+      Value i = for_op.getInductionVar();
+      b.create<memref::StoreOp>(loc, zero_floating, segment_count_memref, i);
```

Comment:
Just curious could this for loop be translated to a true `memset` after lowering ?